File: app/assets/javascripts/playmat/Playmat.js
/* global define:false */
/* global Phaser:false */
define([
  'playmat/helpers/PlaymatResizer', 'playmat/helpers/PlaymatEmitter',
  'card/CardAnimation', 'ui/GameResults'
], function (PlaymatResizer, PlaymatEmitter, CardAnimation, GameResults) {
  'use strict';

  var Playmat = function(game, pileSets, layoutString) {
    Phaser.Group.call(this, game, null, 'playmat');
    this.game.add.existing(this);

    this.pileSets = pileSets;
    this.layoutString = layoutString;

    this.resizer = new PlaymatResizer(this);
    this.resizer.refreshLayout();

    this.emitter = new PlaymatEmitter(this);
  };

  Playmat.prototype = Object.create(Phaser.Group.prototype);
  Playmat.prototype.constructor = Playmat;

  Playmat.prototype.addPile = function(pile) {
    var pileLocation = this.layout.locations[pile.id];
    if(pileLocation === undefined) {
      throw 'Cannot find location for pile [' + pile.id + '].';
    }
    pile.x = pileLocation.x * this.game.cardSet.cardWidth;
    pile.y = pileLocation.y * this.game.cardSet.cardHeight;
  };

  Playmat.prototype.win = function(msg) {
    this.game.complete = true;

    CardAnimation.win(this);

    if(!GameResults.initialized()) {
      GameResults.init();
    }
    GameResults.show(msg.result, msg.stats, true);
  };

  Playmat.prototype.lose = function() {
    //alert('You lose.');
  };

  return Playmat;
});

File: client/src/main/scala/phaser/playmat/Playmat.scala
package phaser.playmat

import com.definitelyscala.phaser.Group
import models.{GameLost, GameWon}
import models.pile.set.PileSet
import phaser.PhaserGame
import phaser.card.CardSprite
import phaser.pile.PileGroup
import utils.NullUtils

import scala.scalajs.js.annotation.ScalaJSDefined

@ScalaJSDefined
class Playmat(val phaser: PhaserGame, val pileSets: Seq[PileSet], val layoutString: String) extends Group(
  game = phaser, parent = NullUtils.inst, name = "playmat"
) {
  var w: Double = 0.0
  var h: Double = 0.0

  private[this] var cards = Map.empty[Int, (CardSprite, PileGroup, Int)]
  def getCardSprite(id: Int) = cards(id)._1
  def getCards = cards.values.toSeq
  def setCards(c: Map[Int, (CardSprite, PileGroup, Int)]) = cards = c

  val emitter = new PlaymatEmitter(this)

  val resizer = new PlaymatResizer(this)
  resizer.refreshLayout(animate = false)

  phaser.add.existing(this)

  private[this] val pileGroups = collection.mutable.HashMap.empty[String, PileGroup]

  def getPileGroups = pileGroups
  def getPileGroup(id: String) = pileGroups.getOrElse(id, throw new IllegalStateException(s"Pile group [$id] not found."))

  def addPileGroup(pileGroup: PileGroup) = {
    pileGroups(pileGroup.id) = pileGroup
    val pileLocation = resizer.layout._3.getOrElse(pileGroup.id, throw new IllegalStateException(s"Missing layout for [${pileGroup.id}]."))
    pileGroup.x = pileLocation._1 * phaser.getSettings.cardSet.w
    pileGroup.y = pileLocation._2 * phaser.getSettings.cardSet.h
  }

  def initialMovesComplete() = pileGroups.foreach(_._2.cards.foreach(_.bringToTop()))

  def win(gw: GameWon) = utils.Logging.info("Win!")
  def lose(gl: GameLost) = utils.Logging.info("Lose!")
}
